### DISCUSSION TITLE
Double validation

### DISCUSSION THREAD

**Discussion by aruznieto** [pymeasure/pymeasure#786] [Q&A]
I am adding an instrument and I would like to know If it is possible a validation for 2 parameters in the same control.

Thanks

**Comment by BenediktBurger** [comment]
If the control requires 2 parameters, they are, effectively, a tuple and you can validate, whether the tuple satisfies some condition. Maybe you have to write your own validator, which does that, but it is possible and not difficult.

If you set, for example, `inst.some_control = 5, 8.9`, the validator gets the tuple `(5, 8.9)` as `value` and compares it to the `values` defined in the property or just according to rules hard coded in the validator.
A validator accepts two parameters: `value` and `values` and has to return a modified `value` or raise an exception.

Don't hesitate to ask, if you require more information.

**Reply by aruznieto** [reply]
Hi, I am trying to doing it, but I do not know why is not working. I try to explain it:

The instrument accepts to configure the central frequency passing the units (Hz, MHz, kHz, GHz). This is my validator where _2 is units and _1 is the number (frequency)

```python
def double_validation_unit_and_freq(value, values):
    pass_discrete = None
    pass_range = None

    value_2 = value[0]
    value_1 = value[1]
    values_2 = values[0]
    values_1 = values[1]

    if value_1 in values_1:
        pass_discrete = value_1
    else:
        raise ValueError('Value of {:s} is not in the discrete set {:s}'.format(
            value_1, values_1
        ))

    if min(values_2) <= value_2 <= max(values_2):
        pass_range = value_2
    else:
        raise ValueError('Value of {:g} is not in range [{:g},{:g}]'.format(
            value_2, min(values_2), max(values_2)
        ))

    return [pass_discrete, pass_range]
```

The control is:
```python

    UNITSFREQ = ['Hz', 'kHz', 'MHz' , 'GHz']

    frequency_center = Instrument.control(
        "FREQuency:CENTer?", "FREQuency:CENTer %g %s",
        "Sets the center frequency in Hz",
        validator=double_validation_unit_and_freq,
        values=[[-99999999995, 299999999995], UNITSFREQ]
    )
```

And the test:

```python
def test_freq_conf():
    with expected_protocol(
            AnritsuMS2090A,
            [(b"FREQuency:CENTer?", (2, 'MHz'))],
            ) as instr:
        assert instr.frequency_center == (2, 'MHz')
```

Test throws me: `TypeError: 'str' object cannot be interpreted as an integer`

**Reply by BenediktBurger** [reply]
The control and validator seem right, the test is faulty: The communication pairs contain the communication in bytes or string, so it should be:
~~~python
[(b"FREQuency:CENTer?", "2 MHz")],
~~~
instead (if the device sends the string "2 MHz").

**Reply by aruznieto** [reply]
> The control and validator seem right, the test is faulty: The communication pairs contain the communication in bytes or string, so it should be:
> 
> ~~~python
> 
> [(b"FREQuency:CENTer?", "2 MHz")],
> 
> ~~~
> 
> instead (if the device sends the string "2 MHz").

Mmm, thank you! I will try it later. Can I add the new instrument to pymeasure, and update it? There are a lot of instructions to control it, and now I only add the most important.

**Reply by BenediktBurger** [reply]
Yes, you can make a first version, if it contains everything you need to use the instrument in a useful manner (I guess these are the "most important" ones you mentioned).
Later you (or somebody else) can implement more controls for more special use cases.

**Reply by aruznieto** [reply]
Ok. Thanks, and how can I test the validator? I am trying to write wrong units, but test does not throw any error.

**Reply by BenediktBurger** [reply]
Have you worked with [pytest](https://docs.pytest.org/)?

A simple example:
~~~python
def test_xy_validator_valid():
    assert xy_validator([5, 'MHz'], [[-99999999995, 299999999995], UNITSFREQ]) == 5

def test_xy_validator_invalid():
    with pytest.raises(ValueError, match="some text"):
        # verifies, that a ValueError is raised with "some text" in the error message.
        xy_validator([-1e100, 'MHz'],  [[-99999999995, 299999999995], UNITSFREQ])
~~~

For testing several values, you can use
~~~python
pytest.mark.parametrize("value", ([5, 'MHz'], [100, 'kHz'], ...))
def test_xy_validator_multiple_values(value):
    assert xy_validator(value, [[-99999999995, 299999999995], UNITSFREQ]) == value[0]
~~~
replace the "..." with more valid values for `value`.


Btw. the validator work on **user input**, not on the returned values!
